Content.Server/GameObjects/Components/Explosion/ClusterFlashComponent.cs

```diff
@@ -0,0 +1,180 @@
+        private float _delay;
+
+        /// <summary>
+        ///     Max distance grenades can be thrown.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        private float _throwDistance;
+
+        /// <summary>
+        ///     This is the end.
+        /// </summary>
+        private bool _countDown;
+
+        // I'm suss on this bit
+        async Task<bool> IInteractUsing.InteractUsing(InteractUsingEventArgs args)
+        {
+            if (_grenadesContainer.ContainedEntities.Count >= _maxGrenades || !args.Using.HasComponent<FlashExplosiveComponent>())
+                return false;
+
+            _grenadesContainer.Insert(args.Using);
+            UpdateAppearance();
+            return true;
+        }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(ref _fillPrototype, "fillPrototype", null);
+            serializer.DataField(ref _maxGrenades, "maxGrenadesCount", 4);
+            serializer.DataField(ref _delay, "delay", 1.0f);
+            serializer.DataField(ref _throwDistance, "distance", 3.0f);
+            serializer.DataField(ref _maxViewableGrenades, "maxView", 3);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            _grenadesContainer = ContainerManagerComponent.Ensure<Container>("cluster-flash", Owner);
+
+        }
+
+        protected override void Startup()
+        {
+            base.Startup();
+
+            if (_fillPrototype != null)
+                FillContainer();
+        }
+
+        bool IUse.UseEntity(UseEntityEventArgs eventArgs)
+        {
+            Owner.SpawnTimer((int) (_delay * 1000), () =>
+            {
+                if (Owner.Deleted || _countDown || ((_grenadesContainer.ContainedEntities.Count) <= 0 && (_fillPrototype == null)))
+                    return;
+
+                _countDown = true;
+                var random = IoCManager.Resolve<IRobustRandom>();
+                var delay = 20;
+                var grenadesWasInserted = _grenadesContainer.ContainedEntities.Count;
+                while (TryGetGrenade(out var grenade))
+                {
+                    var segmentAngle = Convert.ToInt32(360 / grenadesWasInserted);
+                    var throwedCount = grenadesWasInserted - _grenadesContainer.ContainedEntities.Count;
+                    var angleMin = segmentAngle * throwedCount;
+                    var angleMax = segmentAngle * (throwedCount + 1);
+                    // Okay ThrowHelper is actually disgusting and so is this
+                    var angle = Angle.FromDegrees(random.Next(angleMin, angleMax));
+                    var distance = (float)random.NextDouble() * _throwDistance;
+                    var target = new EntityCoordinates(Owner.Uid, angle.ToVec().Normalized * distance);
+
+                    grenade.Throw(1f, target, grenade.Transform.Coordinates);
```

Comment:
1. throwedCount could just be an int that is incremented every time.
2. Is there a reason you used Convert.ToInt32 instead of just a regular cast?
3. I'm 90% sure it was mentioned on discord that you should have an initial offset that's consistent between all of the grenades if you want to have a semi-uniform pattern but these angles are all being started in the loop.
4. angleMin and angleMax should be consistent as well (maybe or maybe not based on number of grenades thrown but definitely not based on our current index).
5. Use random.NextFloat()
6. There's a reason I said the throw code is disgusting, because this won't work if the entity is holding it when it goes off and it doesn't have thrownitemcollision (currently it just explodes in their face). This is due to the fact that when the entity moves it will move the grenades with them.
7. IMO grenades shouldn't explode on impact (though this is more a problem with throwing code).

Reply:
1. Yes
2. Fixed
3. Fixed?
4. Fixed?
4. There is no random.NextFloat()
6. Ok. Deleted.
7. Fixed!